### ISSUE TITLE
Prefixing a route with @UseBefore(Middleware) will run the middleware for every route in class.

### ISSUE THREAD

**Issue by NoNameProvided** [pleerock/routing-controllers#131]
Title says all. When I have a class with multiple routes (`@Get`, `@Put` etc) the `@UseBefore(AuthenticatedMiddleware)` will run for every route

```ts
@JsonController('/auth')
export class AuthRoute {

  constructor(
    @OrmRepository(User) private repository: Repository<User>,
  ) { }

  public createPasswordHash(password: string): Promise<string> {
    return bcrypt.hash(password, 11);
  }

  @Post('/login')
  async login( @BodyParam('email') email: string, @BodyParam('password') password: string) {
    // Login the user
  }

  @UseBefore(AuthenticatedMiddleware)
  @Get('/me')
  public async me( @Req() req: Request) {
    // load user data
  }
```

When I call `/auth/login` the `AuthenticatedMiddleware` should not run but it does.

**Comment by NoNameProvided** [comment]
Hi @pleerock I just created a reproducible minimal use case for this bug to make it easier to track down.  You can find it here: 
[NoNameProvided/routing-controller-bugs#broken-middleware-execution-per-route](/NoNameProvided/routing-controller-bugs/tree/master/broken-middleware-execution-per-route)

**Comment by NoNameProvided** [comment]
Ok so it seems in runs for every route in the app, not only in the class. The root of this is most likely in the `registerMiddleware` function in [ExpressDriver.ts#L44](https://github.com/pleerock/routing-controllers/blob/c9e28df9b666b63d6970bedbb73d735ff3036d47/src/driver/express/ExpressDriver.ts#L44) which simply register every driver as a global one.

**Comment by 19majkel94** [comment]
Everything works correct. You have declared your `AuthenticatedMiddleware` as `@Middleware({ type: "before" })` which means that it's a global before middleware, according to changes in 0.7.0:
> - per-controller and per-action middlewares used in @UseBefore and @UseAfter now should not be marked with @Middleware decorator
> - @MiddlewareGlobalBefore() and @MiddlewareGlobalAfter() were removed and instead new signatures should be used: @Middleware({ type: "before" }) and @Middleware({ type: "after" })

You just have to remove the decorator from `AuthenticatedMiddleware` and it will work only with routes decorated with `@UseBefore`, not as a global before middleware like now :wink:

**Comment by 19majkel94** [comment]
@pleerock I see that the new `@Middleware` decorator usage is kinda confusing. Maybe we should switch to registering global middlewares in `createExpressApp` and remove this decorator in the end? Like now with TypeORM we have a `@Entity` decorator but we still have to register it. Also `@Controller` in routing-controllers has the same behavior.

**Comment by NoNameProvided** [comment]
Damn, I have read the notes, but seems  I had forget that! Thanks. 

On the other side, this should be more clearly mentioned in the docs, not? Something like:

> ## Using middlewares
>
>You can use any existing express / koa middleware or create your own. You can create your own global middlewares with the `@Middleware` decorator, to use existing middlewares on controllers or routes you have to use the `@UseBefore` and `@UseAfter` decorators. You can create custom middlewares for specific controllers our routes with implementing the `ExpressMiddlewareInterface ` or `KoaMiddlewareInterface` interface.

**Comment by NoNameProvided** [comment]
>  Maybe we should switch to registering global middlewares in createExpressApp and remove this decorator in the end? 

I think thats a good idea, it looks weird when I have to import them before my code, but then I doesn't have to do anything with them.

It would be nice to discuss the support of objects instead of an arrays in `createExpressApp`. Because when importing everything from a barrel file it will be an object, containing every middleware with the key of its name. With this approach we would be able to write it like this:

```ts
import * as middlewares from "./middlewares";

this.express = createExpressServer({
      routePrefix: '',
      defaultErrorHandler: false,
      middlewares: middlewares,
    });
```

With the current approach now I am doing this: 

```ts
import * as middlewares from "./middlewares";
middlewaresArray = Object.keys(middlewares).map(key => (<any>middlewares)[key]);

this.express = createExpressServer({
      routePrefix: '',
      defaultErrorHandler: false,
      middlewares: [...middlewaresArray],
    });
```

Not so elegant, but better than listing every entity. However a strong point for this approach is that it easily composable. eg I can write `middlewares: [...middlewaresArray, ...otherMiddleWareSource, 3rdPartyAuthMiddleware],`

**Comment by 19majkel94** [comment]
> It would be nice to discuss the support of objects instead of an arrays in createExpressApp. 

Nope, that's really bad idea. If you want to load middlewares from folder, you can do it in this way:
```ts
this.express = createExpressServer({
    middlewares: [ __dirname + "/../middlewares/*.js" ],
});
```
Or the preffered, explicit way:
```ts
this.express = createExpressServer({
    middlewares: [
        AuthorizationMiddleware,
        CustomErrorMiddleware,
    ]
});
```

And I'm talking about deprecating `@Middleware` decorator in favour of configuration on bootstrap:
```ts
this.express = createExpressServer({
    middlewares: {
        before: [
            AuthorizationMiddleware,
        ],
        after: [
            LoggingMiddleware,
        ],
        error: [
            CustomErrorMiddleware,
        ]
    }
});
```
Or maybe more meaningfull:
```ts
this.express = createExpressServer({
    middlewares: {
        before: [
            AuthorizationMiddleware,
        ],
        after: [
            LoggingMiddleware,
        ],
    },
    errorHandlers: [
        MyErrorResponseHandler,
    ]
});
```

**Comment by NoNameProvided** [comment]
I have checked without the `Middleware` decorator, but it seems throwing errors is not supported in `ExpressMiddlewareInterface`? When I throw an error I get a ` UnhandledPromiseRejectionWarning`, I would expect to be able to throw errors in `ExpressMiddlewareInterface` which will trigger my error handler. (The same way as throwing errors in the controllers will trigger the error handler.)

**Comment by NoNameProvided** [comment]
Created a PR for 👆  this one: #138

**Comment by 19majkel94** [comment]
> But without decorator it won't be possible to determine if its middleware or just regular class.

Yeah, but we could then just let simple `@Middleware()` decorator with no param. Then we could just load from the directories:
```ts
this.express = createExpressServer({
    middlewares: {
        before: [path.join(__dirname, "../middlewares/before/**/*.js")],
        after: [path.join(__dirname, "../middlewares/after/**/*.js")],
        error: [path.join(__dirname, "../middlewares/error/**/*.js")],
    }
});
```
So just instead of weird options and rules no/with decorator on middleware class, we unify all middlewares to simple `@Middleware()` and let users choose the type based on explicit server config. It could be an array of classes or array of glob path strings.

**Comment by pleerock** [comment]
This syntax looks good, and I dislike having `{  after: true }` in decorator as well. However this approach will require users to separate their middlewares on directory level, e.g. `/middleware/after` and `/middleware/before`. Most people have simple middlewares they want to run only before and/or after and having such complex structure just to separate two middlewares can be overkill for them.

**Comment by pleerock** [comment]
I'll close it because PR is merged, @19majkel94 lets continue our discussion on your topic here, or in a separate issue.

**Comment by 19majkel94** [comment]
> Most people have simple middlewares they want to run only before and/or after and having such complex structure just to separate two middlewares can be overkill for them.

1) `before and/or after`
So it's possible to decorate one middleware class with two decorators? One with after, one with before? Or a middleware can be both before and after (I don't see an use case of before-after middleware btw).

2) `having such complex structure just to separate two middlewares can be overkill for them.`
```ts
this.express = createExpressServer({
    middlewares: {
        before: [path.join(__dirname, "../middlewares/auth.js")],
        after: [path.join(__dirname, "../middlewares/logging.js")],
        error: [path.join(__dirname, "../middlewares/error/error-handler.js")],
    }
});
```
And if you don't like listing every middleware file and folder structure:
```ts
this.express = createExpressServer({
    middlewares: {
        before: [path.join(__dirname, "../middlewares/before/*.before.js")],
        after: [path.join(__dirname, "../middlewares/after/*.after.js")],
        error: [path.join(__dirname, "../middlewares/error/*.error.js")],
    }
});
```
It's an elastic solution, you can choose your own way. `{ after: true }` isn't bad but user's are confused right now and register local middleware, etc.

**Comment by pleerock** [comment]
for bigger projects with lot of modules first approach won't be good. Second approach forces user to apply some naming convention which is not really necessary. Even I don't like `{ after: true }` syntax - its not that bad. Especially if we admit that after-middlewares aren't often scenarios. Users are confused only because before it was required for all middlewares to have `@Middleware()` decorator, right now only global middlewares require it.

**Comment by NoNameProvided** [comment]
I like the `*.before`, `*.after`, `*.error` naming convention, I found it helpful generally, as it tells you exactly what is in the file. We already use this all the time, so eg services in our projects also has a `*.services` name. So the second solutions seems ok for me. 

Btw totally unrelated question, but if you read middlewares from a directory which order will they be executed? The same as it was read from the disk? In this case I dont even see the benefit of reading middlewares from the disk, because usually the order is different than the alphabetical order.

**Comment by 19majkel94** [comment]
> Users are confused only because before it was required for all middlewares to have @Middleware() decorator, right now only global middlewares require it.

Yeah, so they are trying to decorator local middleware with `@Middleware() ` decorator, which force them to specify whether it's before or after, which implicitly set it as a global middleware. I other words: they want to create a local middleware, read the doc, remember the `@Middleware()` decorator, use it, it force them to specify before/after which as I said implicitly set it as a global middleware.

So maybe if it's only required for global middleware, we should go back to `@GlobalMiddleware` decorator naming, which is self decribing? Or we can add `type: local` option to the decorator (I don't like it, it's confusing too): 
```ts
 export function Middleware(options: { type: "after"|"before"|"local", priority?: number }): Function 
```

And btw, this approach:
```ts
this.express = createExpressServer({
    middlewares: {
        before: [
            AuthorizationMiddleware,
        ],
        after: [
            LoggingMiddleware,
        ],
        error: [
            CustomErrorMiddleware,
        ]
    }
});
```
Has one big advantage - we specify the order of middleware, just like in pure express app. No need to change `priority` option in 10 middleware classes to add new middleware to the begin of the chain.

**Comment by NoNameProvided** [comment]
> Has one big advantage - we specify the order of middleware, just like in pure express app.

Yes, I agree, I would even say dropping support for loading via route should be dropped. Explicitly over laziness sometimes can save a lot of times. And if the user really dont care, there is a workaround: 

```ts
import * as Middlewares from './middlewares';

this.express = createExpressServer({
    middlewares: {
        before: Object.keys(Middlewares).map(key => Middlewares[key]),
    }
});
```

**Comment by 19majkel94** [comment]
> `Object.keys(Middlewares).map(key => Middlewares[key])`

Yeah, but we are going back to the problem - the order of middleware is undetermined, object property iteration order is not guaranteed. So we again need a decorator to configure it.

**Comment by NoNameProvided** [comment]
> Yeah, but we are going back to the problem 

I would consider `Object.keys(Middlewares).map(key => Middlewares[key])` a hack to use if u really dont care about the order (Is there even any scenario when you really dont care?), but I would consider it a bad practice what should be avoided. (Just like loading middlewares from path.)

I would not promote this in the docs. While loading middlewares from folders is promoted in the docs currently while still have the same out of order loading issue.

> So we again need a decorator to configure it.

I would never used the priority tag, it is a way more harder to track the order of components if you have to open 10+ files to see which has the highest priority.
